File: crypto-tests/benches/encryption.rs
use criterion::{criterion_group, criterion_main, Bencher, BenchmarkId, Criterion};
use crypto_tests::{decrypt_payload, encrypt_hdr, encrypt_payload};
use crypto_tests::{encrypt_packet, MyNonce};
use ring::aead::quic::{HeaderProtectionKey, AES_128};
use ring::aead::SealingKey;
use ring::aead::UnboundKey;
use ring::aead::{BoundKey, OpeningKey};
use ring::aead::{AES_128_GCM, NONCE_LEN};

pub fn criterion_benchmark(c: &mut Criterion) {
    let key_bytes = [
        10, 141, 102, 148, 37, 119, 128, 179, 47, 14, 68, 0, 205, 28, 26, 149,
    ];
    let nonce = MyNonce {
        nonce: [0; NONCE_LEN],
    };
    let algorithm = &AES_128_GCM;
    let unbound_key = UnboundKey::new(&algorithm, &key_bytes).unwrap();
    let mut key = SealingKey::<MyNonce>::new(unbound_key, nonce);
    let mut header = vec![
        0x57, 0x25, 0xe7, 0x4f, 0x2d, 0x27, 0x5d, 0x12, 0x8b, 0x37, 0xb0, 0x47, 0x04, 0x16, 0x08,
        0xa1, 0x84, 0x23, 0x65, 0xdb, 0xfa, 0xe7,
    ];
    let hp_key_bytes = [
        0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e,
        0x0f,
    ];
    let mut hp_key = HeaderProtectionKey::new(&AES_128, &hp_key_bytes).unwrap();
    let mut group = c.benchmark_group("payload vs packet encryption");
    for payload in generate_data().iter_mut() {
        group.bench_with_input(
            BenchmarkId::new("payload encryption", payload.len()),
            payload,
            |b, payload| {
                b.iter(|| {
                    encrypt_payload(payload.clone().as_mut(), header.clone().as_mut(), &mut key)
                })
            },
        );
        group.bench_with_input(
            BenchmarkId::new("packet encryption", payload.len()),
            payload,
            |b, payload| {
                b.iter(|| {
                    encrypt_packet(
                        header.clone().as_mut(),
                        payload.clone().as_mut(),
                        &mut hp_key,
                        &mut key,
                    )
                })
            },
        );
    }
    group.finish();

    let mut group2 = c.benchmark_group("header vs payload encryption");
    for payload in generate_data().iter_mut() {
        group2.bench_with_input(
            BenchmarkId::new("payload encryption", payload.len()),
            payload,
            |b, payload| {
                b.iter(|| {
                    encrypt_payload(payload.clone().as_mut(), header.clone().as_mut(), &mut key)
                })
            },
        );
        group2.bench_with_input(
            BenchmarkId::new("header encryption", payload.len()),
            payload,
            |b, payload| b.iter(|| encrypt_hdr(header.clone().as_mut(), payload, &mut hp_key)),
        );
    }
    group2.finish();

    let o_nonce = MyNonce {
        nonce: [0; NONCE_LEN],
    };
    let o_unbound_key = UnboundKey::new(&algorithm, &key_bytes).unwrap();
    let mut o_key = OpeningKey::<MyNonce>::new(o_unbound_key, o_nonce);
    let mut group3 = c.benchmark_group("payload encryption vs decryption");
    let mut payloads = generate_data();
    let mut encrypted_payloads = payloads.clone();
    encrypt_payloads(&mut key, &mut header, &mut encrypted_payloads);
    for i in 0..payloads.len() {
        let payload = payloads.get(i).unwrap();
        let encrypted_payload = encrypted_payloads.get(i).unwrap();
        group3.bench_with_input(
            BenchmarkId::new("payload encryption", payload.len()),
            payload,
            |b, payload| {
                b.iter(|| {
                    encrypt_payload(payload.clone().as_mut(), header.clone().as_mut(), &mut key)
                })
            },
        );
        group3.bench_with_input(
            BenchmarkId::new("payload decryption", payload.len()),
            encrypted_payload,
            |b, payload| {
                b.iter(|| {
                    decrypt_payload(
                        &mut o_key,
                        header.clone().as_mut(),
                        encrypted_payload.clone().as_mut(),
                    )
                })
            },
        );
    }
    group3.finish();
}

fn generate_data() -> Vec<Vec<u8>> {
    vec![
        100, 200, 300, 400, 500, 600, 700, 800, 900, 1000, 1100, 1200, 1300,
    ]
    .into_iter()
    .map(|bytes| (0..bytes).map(|_| rand::random::<u8>()).collect())
    .collect()
}

fn encrypt_payloads(
    key: &mut SealingKey<MyNonce>,
    header: &mut Vec<u8>,
    payloads: &mut Vec<Vec<u8>>,
) {
    payloads
        .into_iter()
        .for_each(|payload| encrypt_payload(payload, header, key));
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);

File: crypto-tests/elixir_ct/benches/encryption.exs
defmodule Benchmark do
  @header <<0x57, 0x25, 0xE7, 0x4F, 0x2D, 0x27, 0x5D, 0x12, 0x8B, 0x37, 0xB0, 0x47, 0x04, 0x16,
            0x08, 0xA1, 0x84, 0x23, 0x65, 0xDB, 0xFA, 0xE7>>
  @key <<10, 141, 102, 148, 37, 119, 128, 179, 47, 14, 68, 0, 205, 28, 26, 149>>
  @hp_key <<0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D,
            0x0E, 0x0F>>

  def run() do
    Benchee.run(
      %{
        "payload_enc" => fn {payload, crypto_state} ->
          ElixirCt.encrypt_payload(crypto_state, @header, payload)
        end,
        "packet_enc" => fn {payload, crypto_state} ->
          ElixirCt.encrypt_pkt(crypto_state, @header, payload)
        end
      },
      before_scenario: fn payload ->
        crypto_state = ElixirCt.init(@key, @hp_key, :aes_128_gcm)
        {payload, crypto_state}
      end,
      inputs: %{
        "100" => ElixirCt.generate_payload(100),
        "200" => ElixirCt.generate_payload(200),
        "300" => ElixirCt.generate_payload(300),
        "400" => ElixirCt.generate_payload(400),
        "500" => ElixirCt.generate_payload(500),
        "600" => ElixirCt.generate_payload(600),
        "700" => ElixirCt.generate_payload(700),
        "800" => ElixirCt.generate_payload(800),
        "900" => ElixirCt.generate_payload(900),
        "1000" => ElixirCt.generate_payload(1000),
        "1100" => ElixirCt.generate_payload(1100),
        "1200" => ElixirCt.generate_payload(1200),
        "1300" => ElixirCt.generate_payload(1300)
      },
      formatters: [
        Benchee.Formatters.Console,
        {Benchee.Formatters.HTML, file: "benchee/report.html"}
      ]
    )
  end
end

Benchmark.run()
